hazelcast-client/src/main/java/com/hazelcast/client/util/ClientDelegatingFuture.java

```diff
@@ -189,87 +177,62 @@ public V join() {
-            // if "shouldDeserializeData" flag is not true in any of registered "DelegatingExecutionCallback".
-            // So, be sure that value is deserialized before returning to caller.
-            deserializedValue = serializationService.toObject(response);
-            return deserializedValue;
+        ClientMessage tempMessage = ClientMessage.createForDecode(clientMessage.buffer(), 0);
+        Object decodedResponse = clientMessageDecoder.decodeClientMessage(tempMessage);
+        if (deserialize) {
+            return serializationService.toObject(decodedResponse);
         }
-
-        return response;
-    }
-
-    private Object resolveMessageToValue(ClientMessage message) {
-        return clientMessageDecoder.decodeClientMessage(message);
+        return decodedResponse;
     }
 
     protected void setError(Throwable error) {
-        this.error = error;
-    }
-
-    protected void setDone() {
-        this.done = true;
+        responseMessage.compareAndSet(NULL_OBJECT, error);
     }
 
     protected ClientInvocationFuture getFuture() {
         return future;
     }
 
-    private boolean isResponseSet() {
-        return response != mutex;
-    }
-
     class DelegatingExecutionCallback<T> implements ExecutionCallback<ClientMessage> {
 
         private final ExecutionCallback<T> callback;
-        private final boolean shouldDeserializeData;
+        private final boolean deserialize;
 
-        DelegatingExecutionCallback(ExecutionCallback<T> callback, boolean shouldDeserializeData) {
+        DelegatingExecutionCallback(ExecutionCallback<T> callback, boolean deserialize) {
             this.callback = callback;
-            this.shouldDeserializeData = shouldDeserializeData;
+            this.deserialize = deserialize;
         }
 
         @Override
         public void onResponse(ClientMessage message) {
-            if (!done || !isResponseSet()) {
-                synchronized (mutex) {
-                    if (!done || !isResponseSet()) {
-                        response = resolveMessageToValue(message);
-                        if (shouldDeserializeData && deserializedValue == null) {
-                            deserializedValue = serializationService.toObject(response);
-                        }
-                        done = true;
-                    }
-                }
-            }
-            if (shouldDeserializeData) {
-                callback.onResponse((T) deserializedValue);
-            } else {
-                callback.onResponse((T) response);
-            }
+            complete(message);
         }
 
         @Override
         public void onFailure(Throwable t) {
-            if (!done) {
-                synchronized (mutex) {
-                    if (!done) {
-                        error = t;
-                        done = true;
-                    }
-                }
+            complete(t);
+        }
+
+        private void complete(Object remoteResponse) {
```

Comment:
I would not call this method 'complete'. Complete has a special meaning on the completable future.